modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,452 @@
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TASK;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_COMMIT;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_ROLLBACK;
+
+/**
+ * Performance statistics writer based on logging to a file.
+ * <p>
+ * Each node collects statistics to a file placed under {@link #PERFORMANCE_STAT_DIR}.
+ * <p>
+ * <b>Note:</b> Start again will erase previous performance statistics files.
+ * <p>
+ * To iterate over records use {@link FilePerformanceStatisticsReader}.
+ */
+public class FilePerformanceStatisticsWriter {
+    /** Directory to store performance statistics files. Placed under Ignite work directory. */
+    public static final String PERFORMANCE_STAT_DIR = "performanceStatistics";
+
+    /** Default maximum file size in bytes. Performance statistics will be stopped when the size exceeded. */
+    public static final long DFLT_FILE_MAX_SIZE = 32 * U.GB;
+
+    /** Default off heap buffer size in bytes. */
+    public static final int DFLT_BUFFER_SIZE = (int)(32 * U.MB);
+
+    /** Default minimal batch size to flush in bytes. */
+    public static final int DFLT_FLUSH_SIZE = (int)(8 * U.MB);
+
+    /** Factory to provide I/O interface. */
+    private final FileIOFactory fileIoFactory = new RandomAccessFileIOFactory();
+
+    /** Performance statistics file writer worker. */
+    private final FileWriter fileWriter;
+
+    /** Performance statistics file I/O. */
+    private final FileIO fileIo;
+
+    /** File write buffer. */
+    private final SegmentedRingByteBuffer ringByteBuffer;
+
+    /** Count of written to buffer bytes. */
+    private final AtomicInteger writtenToBuffer = new AtomicInteger();
+
+    /** {@code True} if the small buffer warning message logged. */
+    private final AtomicBoolean smallBufLogged = new AtomicBoolean();
+
+    /** {@code True} if worker stopped due to maximum file size reached. */
+    private final AtomicBoolean stopByMaxSize = new AtomicBoolean();
+
+    /** Hashcodes of cached strings. */
+    private final ConcurrentSkipListSet<Integer> cachedStrings = new ConcurrentSkipListSet<>();
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** @param ctx Kernal context. */
+    public FilePerformanceStatisticsWriter(GridKernalContext ctx) throws IgniteCheckedException, IOException {
+        log = ctx.log(getClass());
+
+        File file = statisticsFile(ctx);
+
+        U.delete(file);
+
+        fileIo = fileIoFactory.create(file);
+
+        log.info("Performance statistics file created [file=" + file.getAbsolutePath() + ']');
+
+        ringByteBuffer = new SegmentedRingByteBuffer(DFLT_BUFFER_SIZE, DFLT_FILE_MAX_SIZE,
+            SegmentedRingByteBuffer.BufferMode.DIRECT);
+
+        fileWriter = new FileWriter(ctx, log);
+    }
+
+    /** Starts collecting performance statistics. */
+    public void start() {
```

Comment:
Why do we need an explicit `start` method? We can start `fileWriter` thread from constructor.